neuralmonkey/encoders/cnn_encoder.py

```diff
@@ -76,97 +57,112 @@ def __init__(self,
+        attribute.  There is not dropout between the convolutional layers, by
+        default the activation function is ReLU.
+        """
+        last_layer = self.image_input
+        image_processing_layers = []  # type: List[tf.Tensor]
+
+        with tf.variable_scope("convolutions"):
+            for i, (filter_size,
+                    n_filters,
+                    pool_size) in enumerate(self.convolutions):
+                with tf.variable_scope("cnn_layer_{}".format(i)):
+                    last_layer = conv2d(last_layer, n_filters, filter_size)
+                    image_processing_layers.append(last_layer)
+
+                    if pool_size:
+                        last_layer = max_pool2d(last_layer, pool_size)
+                        image_processing_layers.append(last_layer)
+
+        return image_processing_layers
+
+    @tensor
+    def states(self):
+        # pylint: disable=unsubscriptable-object
+        return self.image_processing_layers[-1]
+        # pylint: enable=unsubscriptable-object
+
+    @tensor
+    def encoded(self) -> tf.Tensor:
+        """Output vector of the CNN.
+
+        If there are specified some fully connected layers, there are applied
+        on top of the last convolutional map. Dropout is applied between all
+        layers, default activation function is ReLU. There are only projection
+        layers, no softmax is applied.
+
+        If there is fully_connected layer specified, average-pooled last
+        convolutional map is used as a vector output.
+        """
+        # pylint: disable=no-member
+        last_height, last_width, last_n_channels = [
+            s.value for s in self.states.get_shape()[1:]]
+        # pylint: enable=no-member
+
+        if self.fully_connected is None:
+            # we average out by the image size -> shape is number
+            # channels from the last convolution
+            encoded = tf.reduce_mean(self.states, [1, 2])
+            assert_shape(encoded, [None, self.convolutions[-1][1]])
+            return encoded
+
+        states_flat = tf.reshape(
+            self.states,
+            [-1, last_width * last_height * last_n_channels])
+        return multilayer_projection(
+            states_flat, self.fully_connected,
+            activation=tf.nn.relu,
+            dropout_keep_prob=self.dropout_keep_prob,
+            train_mode=self.train_mode)
+
+    @tensor
     def _attention_tensor(self) -> tf.Tensor:
-        return self.__attention_tensor
-
-    @property
+        # pylint: disable=no-member
+        last_height, last_width, last_n_channels = [
+            s.value for s in self.states.get_shape()[1:]]
+        # pylint: enable=no-member
+        return tf.reshape(
+            self.states,
+            [-1, last_width * last_height, last_n_channels])
+
+    @tensor
     def _attention_mask(self) -> tf.Tensor:
-        return self.__attention_mask
+        return tf.ones(tf.shape(self._attention_tensor)[:2])
```

Comment:
A co jinýho? Prostě chci dělat attention přes celej ten obrázek. Časem bych to vytunil, že image reader bude jetě vracet masku, co si dopadoval a co bylo původně v tom obrázku, ale teď bych to nechal tak, že se prostě pojede přes celý obrázek.